SpringBox/pair_interactions.py

```diff
@@ -2,57 +2,75 @@
-def spring_forces(acc,pXs , Dij, _config):
+def spring_forces(acc, pXs, Dij, Drxns, _config):
     rhs = np.zeros_like(pXs)
-    n_part=_config['n_part']
-    k=_config['spring_k']
-    r0=_config['spring_r0']
-    Iij = Dij * np.outer(acc,acc)
-    Iij = (Iij>_config['spring_lower_cutoff']) * (Iij<_config['spring_cutoff'])
+    n_part = _config['n_part']
+    k = _config['spring_k']
+    r0 = _config['spring_r0']
+    Iij = Dij * np.outer(acc, acc)
+    Iij = (Iij > _config['spring_lower_cutoff']) * (Iij < _config['spring_cutoff'])
     for i in range(n_part):
-        for j in range(i+1,n_part):
-            if Iij[i,j]!=0:
-                rhs[i] += -k*((Dij[i,j]-r0)/Dij[i,j])*(pXs[i]-pXs[j])
-                rhs[j] += +k*((Dij[i,j]-r0)/Dij[i,j])*(pXs[i]-pXs[j])
+        for j in range(i + 1, n_part):
+            if Iij[i, j] != 0:
+                rhs[i] += -k * ((Dij[i, j] - r0) / Dij[i, j]) * (Drxns[i][j])
+                rhs[j] += +k * ((Dij[i, j] - r0) / Dij[i, j]) * (Drxns[i][j])
     return rhs
+
+
 @numba.jit
-def LJ_forces(acc,pXs , Dij, _config):
-    n_part=_config['n_part']
+def LJ_forces(acc, pXs, Dij, _config):
+    n_part = _config['n_part']
     rhs = np.zeros_like(pXs)
-    eps=_config['LJ_eps']
-    r0=_config['LJ_r0']
-    r0_6=r0**6
-    LJ_pre = 12*eps*r0_6
-    Dij6 = Dij**6
-    np.clip(Dij6, 1e-6, None, out = Dij6)
-    Iij = (Dij<_config['LJ_cutoff']) * np.outer(acc,acc)
+    eps = _config['LJ_eps']
+    r0 = _config['LJ_r0']
+    r0_6 = r0 ** 6
+    LJ_pre = 12 * eps * r0_6
+    Dij6 = Dij ** 6
+    np.clip(Dij6, 1e-6, None, out=Dij6)
+    Iij = (Dij < _config['LJ_cutoff']) * np.outer(acc, acc)
     for i in range(n_part):
-        for j in range(i+1,n_part):
-            if Iij[i,j]!=0:
-                rhs[i] += -LJ_pre*(pXs[i]-pXs[j])*((Dij6[i,j]-r0_6)/(Dij[i,j]*Dij6[i,j]))
-                rhs[j] += +LJ_pre*(pXs[i]-pXs[j])*((Dij6[i,j]-r0_6)/(Dij[i,j]*Dij6[i,j]))
+        for j in range(i + 1, n_part):
+            if Iij[i, j] != 0:
+                rhs[i] += -LJ_pre * (pXs[i] - pXs[j]) * ((Dij6[i, j] - r0_6) / (Dij[i, j] * Dij6[i, j]))
+                rhs[j] += +LJ_pre * (pXs[i] - pXs[j]) * ((Dij6[i, j] - r0_6) / (Dij[i, j] * Dij6[i, j]))
     return rhs
 
+
+@numba.jit
+def periodic_dist(u, v, length):
+    x_diff = min(abs(u[0] - v[0]), 2 * length - abs(u[0] - v[0]))
+    y_diff = min(abs(u[1] - v[1]), 2 * length - abs(u[1] - v[1]))
+    return np.sqrt(x_diff ** 2 + y_diff ** 2)
+
+
+def vectors(pXs, L):
+    diffs = np.array([np.subtract.outer(p, p) for p in pXs.T]).T
+    return np.remainder(diffs + L, 2 * L) - L
+
+
 @numba.jit
 def RHS(pXs, prv_acc, activation_fn, _config):
     rhs = np.zeros_like(pXs)
     acc = active_particles(pXs, prv_acc, activation_fn, _config)
-    Dij = squareform(pdist(pXs))
-    np.clip(Dij, 1e-6, None, out = Dij)
-    assert(_config['spring_k']>0 or _config['LJ_eps']>0)
+    length = _config['L']
+    Dij = squareform(pdist(pXs, metric=periodic_dist, length=length))
```

Comment:
Probably here is a better place to make the check for periodic/non-periodic. Since it affects both Dij and Drxns.